File: subprojects/base-services/src/main/java/org/gradle/internal/xml/XmlFactories.java
/*
 * Copyright 2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gradle.internal.xml;

import org.gradle.api.NonNullApi;
import org.xml.sax.SAXNotRecognizedException;
import org.xml.sax.SAXNotSupportedException;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParserFactory;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerFactory;
import javax.xml.xpath.XPathFactory;
import javax.xml.xpath.XPathFactoryConfigurationException;

/**
 * Factories for javax.xml.
 */
@NonNullApi
public final class XmlFactories {

    public static DocumentBuilderFactory newDocumentBuilderFactory() {
        try {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            return dbf;
        } catch (ParserConfigurationException e) {
            throw new RuntimeException("Unable to create secure DocumentBuilderFactory", e);
        }
    }

    public static SAXParserFactory newSAXParserFactory() {
        try {
            SAXParserFactory spf = SAXParserFactory.newInstance();
            spf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            return spf;
        } catch (ParserConfigurationException e) {
            throw new RuntimeException("Unable to create secure SAXParserFactory", e);
        } catch (SAXNotSupportedException e) {
            throw new RuntimeException("Unable to create secure SAXParserFactory", e);
        } catch (SAXNotRecognizedException e) {
            throw new RuntimeException("Unable to create secure SAXParserFactory", e);
        }
    }

    public static XPathFactory newXPathFactory() {
        try {
            XPathFactory xpf = XPathFactory.newInstance();
            xpf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            return xpf;
        } catch (XPathFactoryConfigurationException e) {
            throw new RuntimeException("Unable to create secure XPathFactory", e);
        }
    }

    public static TransformerFactory newTransformerFactory() {
        try {
            TransformerFactory tf = TransformerFactory.newInstance();
            tf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            return tf;
        } catch (TransformerConfigurationException e) {
            throw new RuntimeException("Unable to create secure TransformerFactory", e);
        }
    }

    private XmlFactories() {}
}

File: build-logic-commons/basics/src/main/kotlin/gradlebuild/basics/XmlFactories.kt
/*
 * Copyright 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gradlebuild.basics

import javax.xml.XMLConstants
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory


/**
 * Creates a [DocumentBuilder] that does not resolve external entities.
 *
 * Build logic cannot use `org.gradle.internal.xml.XmlFactories`, since that lives in the
 * distribution this build produces, so this is the equivalent for build logic.
 */
fun createSecureDocumentBuilder(): DocumentBuilder =
    DocumentBuilderFactory.newInstance().apply {
        setFeature("http://apache.org/xml/features/disallow-doctype-decl", true)
        setFeature("http://xml.org/sax/features/external-general-entities", false)
        setFeature("http://xml.org/sax/features/external-parameter-entities", false)
        setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
        isExpandEntityReferences = false
        setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true)
        setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
        setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "")
    }.newDocumentBuilder()
